[iPad] cbssports.com: Media controls on videos do not respond to trackpad clicks

whsieh opened the pull request:
#### 44b6bb811936b2bd9af8c97c6fbe5a5c4ceb68d6
<pre>
[iPad] cbssports.com: Media controls on videos do not respond to trackpad clicks
<a href="https://bugs.webkit.org/show_bug.cgi?id=282794">https://bugs.webkit.org/show_bug.cgi?id=282794</a>
<a href="https://rdar.apple.com/84667808">rdar://84667808</a>

Reviewed by Abrar Rahman Protyasha.

When watching videos on cbssports.com, it&apos;s not currently possible to interact with the media
player&apos;s custom controls using a trackpad on iPad. This is because the website itself explicitly
prevents click events from triggering any actions on the media player, in the case where touch
capabilities are supported (i.e. `this.isTouch` is set):

```
onClick(e)
{
    this.isTouch || this.disabled ? Dt(e) : this.execAction()
}
onTouchStart(e)
{
    Dt(e),
    this.touchStartStamp = e.timeStamp
}
onTouchEnd(e)
{
    !this.disabled &amp;&amp; this.touchStartStamp &amp;&amp; e.timeStamp - this.touchStartStamp &lt;= this.TAP_THRESHOLD &amp;&amp; (this.touchStartStamp = null, this.execAction())
}
```

To make this Just Work, this patch add a site-specific hack that triggers only when clicking on
specific media controls on cbssports.com (precisely, elements with tag name `AVIA-BUTTON`). If the
click is swallowed (i.e. prevented default), we call out into the client layer to simulate a pair of
`touchstart` / `touchend` events, which results in the control being &quot;clicked&quot;.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::didSwallowClickEvent):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::swallowAnyClickEvent):
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldSynthesizeTouchEventsAfterNonSyntheticClick const):
* Source/WebCore/page/Quirks.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::didSwallowClickEvent):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::didSwallowClickEvent):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::didSwallowClickEvent):

Add logic to synthesize touchstart/touchend events over the center of the clicked element. See above
for more details.

Canonical link: <a href="https://commits.webkit.org/286338@main">https://commits.webkit.org/286338@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d3d5732109758248e601f5c8cd5d38f6503708c7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75651 "Passed style check") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28502 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80136 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/26915 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77767 "Passed tests") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2939 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59331 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17506 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78718 "Passed tests") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/64981 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39694 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22466 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25244 "Built successfully") | 
| | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22802 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81610 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/2990 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1885 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67563 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3141 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64955 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66870 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16671 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10820 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8967 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/2947 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5769 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/2972 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/3907 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/2979 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
```diff
@@ -5670,6 +5670,36 @@ static VisiblePositionRange constrainRangeToSelection(const VisiblePositionRange
+    if (!document->quirks().shouldSynthesizeTouchEventsAfterNonSyntheticClick(node))
+        return;
+
+    bool isReplaced = false;
+    auto bounds = node.absoluteBoundingRect(&isReplaced);
```
Sad that we have to make this accommodation... maybe someone should add this to `Node`

```cpp
LayoutRect Node::absoluteBoundingRect()
{
    bool unusedIsReplaced;
    return absoluteBoundingRect(&unusedIsReplaced);
}
```

whsieh replied:
Yeah, that — or add a null check for the `bool*` outparam (with a `nullptr` default argument).

Might do that in a followup (since a bunch of other existing call sites could adopt it)

whsieh committed:
[iPad] cbssports.com: Media controls on videos do not respond to trackpad clicks
https://bugs.webkit.org/show_bug.cgi?id=282794
rdar://84667808

Reviewed by Abrar Rahman Protyasha.

When watching videos on cbssports.com, it's not currently possible to interact with the media
player's custom controls using a trackpad on iPad. This is because the website itself explicitly
prevents click events from triggering any actions on the media player, in the case where touch
capabilities are supported (i.e. `this.isTouch` is set):

```
onClick(e)
{
    this.isTouch || this.disabled ? Dt(e) : this.execAction()
}
onTouchStart(e)
{
    Dt(e),
    this.touchStartStamp = e.timeStamp
}
onTouchEnd(e)
{
    !this.disabled && this.touchStartStamp && e.timeStamp - this.touchStartStamp <= this.TAP_THRESHOLD && (this.touchStartStamp = null, this.execAction())
}
```

To make this Just Work, this patch add a site-specific hack that triggers only when clicking on
specific media controls on cbssports.com (precisely, elements with tag name `AVIA-BUTTON`). If the
click is swallowed (i.e. prevented default), we call out into the client layer to simulate a pair of
`touchstart` / `touchend` events, which results in the control being "clicked".

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::didSwallowClickEvent):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::swallowAnyClickEvent):
* Source/WebCore/page/EventHandler.h:
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldSynthesizeTouchEventsAfterNonSyntheticClick const):
* Source/WebCore/page/Quirks.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::didSwallowClickEvent):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::didSwallowClickEvent):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::didSwallowClickEvent):

Add logic to synthesize touchstart/touchend events over the center of the clicked element. See above
for more details.

Canonical link: https://commits.webkit.org/286338@main
